File: wk02/src/boyermoore_basic.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

bool boyermoore_basic(const char* pat, const char* txt) {
    int n = strlen(txt);
    int m = strlen(pat);

    int rdict[128] = {0};  // init 0s for array of size 128 for all ASCII chars
    for (int i = 0; i < m; i++)
        rdict[(int)pat[i]] = i;  // assign rightmost ocurrence position in pat

    int x = 0;
    while (x <= n - m) {
        int k = m - 1;
        while (k >= 0 && pat[k] == txt[x + k]) {
            k--;
        }

        if (k == -1) return true;

        int rx = rdict[txt[x + k]];
        x += MAX(k - rx, 1);
    }

    return false;
}

int main() {
    const char* txt = "sphoorythy";
    const char* pat = "thy";
    bool res = boyermoore_basic(pat, txt);
    assert(res == true);

    txt = "xyzsomething";
    pat = "abc";
    res = boyermoore_basic(pat, txt);
    assert(res == false);

    txt = "xxxxxxxxxabc";
    pat = "abc";
    res = boyermoore_basic(pat, txt);
    assert(res == true);

    return 0;
}

File: wk02/src/boyermoore_basic.py
"""
Boyer Moore Basic is a simpler implementation of the full Boyer Moore version.

The only thing is that it ONLY uses the bad character rule. It's simpler and
easier to grasp without good suffix rule and galil's optimisation.

- k is the position in the pattern where a mismatch occurs.

- R(x) is the rightmost occurrence of a character x in the pattern P.
    The thing about this here; unlike the real BM algo, it doesn't change.

- On each mismatch do max(k - R(x), 1) to find out how much to the alignment
  shift by.
"""


def boyermoore_basic(pat: str, txt: str) -> bool:
    n = len(txt)
    m = len(pat)

    rdict = dict([(pat[i], i) for i in range(m)])

    def r(char):
        return rdict.get(char, 0)

    x = 0
    while x <= n - m:
        k = m - 1
        while k >= 0 and pat[k] == txt[x + k]:
            k -= 1

        if k == -1:
            return True

        x += max(k - r(txt[x + k]), 1)

    return False


if __name__ == "__main__":
    txt = "sphoorythy"
    pat = "thy"
    res = boyermoore_basic(pat, txt)
    tru = True
    assert res == tru, f"Expected {tru}, got {res}"

    txt = "xyzsomething"
    pat = "abc"
    res = boyermoore_basic(pat, txt)
    tru = False
    assert res == tru, f"Expected {tru}, got {res}"

    txt = "xxxxxxxxxabc"
    pat = "abc"
    res = boyermoore_basic(pat, txt)
    tru = True
    assert res == tru, f"Expected {tru}, got {res}"
